libdevcore/DBFactory.cpp

```diff
@@ -0,0 +1,151 @@
+#include "FileSystem.h"
+#include "LevelDB.h"
+#include "MemoryDB.h"
+
+namespace dev
+{
+namespace db
+{
+namespace fs = boost::filesystem;
+namespace po = boost::program_options;
+
+auto g_kind = DatabaseKind::LevelDB;
+fs::path g_dbPath = getDataDir();
+
+/// A helper type to build the tabled of DB implementations.
+///
+/// More readable than std::tuple.
+/// Fields are not initialized to allow usage of construction with initializer lists {}.
+struct DBKindTableEntry
+{
+    DatabaseKind kind;
+    const char* name;
+};
+
+/// The table of available DB implementations.
+///
+/// We don't use a map to avoid complex dynamic initialization. This list will never be long,
+/// so linear search only to parse command line arguments is not a problem.
+DBKindTableEntry dbKindsTable[] = {
+    {DatabaseKind::LevelDB, "leveldb"},
+    {DatabaseKind::MemoryDB, "memorydb"},
+};
+
+void setDatabaseKind(const std::string& _name)
+{
+    for (auto& entry : dbKindsTable)
+    {
+        if (_name == entry.name)
+        {
+            g_kind = entry.kind;
+            return;
+        }
+    }
+
+    BOOST_THROW_EXCEPTION(std::system_error(std::make_error_code(std::errc::invalid_argument),
+        "invalid database kind supplied: " + _name));
+}
+
+void setDatabasePath(const std::string& _path)
+{
+    g_dbPath = fs::path(_path);
+}
+
+bool isMemoryDB()
+{
+    return g_kind == DatabaseKind::MemoryDB;
+}
+
+DatabaseKind getDatabaseKind()
+{
+    return g_kind;
+}
+
+fs::path getDatabasePath()
+{
+    return g_dbPath;
+}
+
+po::options_description dbProgramOptions(unsigned _lineLength)
+{
+    // It must be a static object because boost expects const char*.
+    static const std::string description = [] {
+        std::string names;
+        for (auto const& entry : dbKindsTable)
+        {
+            if (!names.empty())
+                names += ", ";
+            names += entry.name;
+        }
+
+        return "Select database implementation. Available options are: " + names + ".";
+    }();
+
+    po::options_description opts("DATABASE OPTIONS", _lineLength);
+    auto add = opts.add_options();
+
+    add("db",
+        po::value<std::string>()->value_name("<name>")->default_value("leveldb")->notifier(
+            setDatabaseKind),
+        description.data());
+
+    add("db-path",
+        po::value<std::string>()
+            ->value_name("<path>")
+            ->default_value(getDataDir().string())
+            ->notifier(setDatabasePath),
+        "Database path (for non-memory database options)\n");
+
+    return opts;
+}
+
+std::unique_ptr<DatabaseFace> DBFactory::create()
+{
+    return create(g_dbPath);
+}
+
+std::unique_ptr<DatabaseFace> DBFactory::create(fs::path _path)
+{
+    return create(g_kind, _path);
+}
+
+std::unique_ptr<DatabaseFace> DBFactory::create(DatabaseKind _kind, fs::path _path)
+{
+    switch (_kind)
+    {
+    case DatabaseKind::LevelDB:
+        return std::unique_ptr<DatabaseFace>(new LevelDB(_path));
+        break;
```

Comment:
`break`s after `return`s are redundant.

Reply:
Yes, thought I needed this to satisfy the compiler, since I don't I'll remove it. Thanks for calling this out!